BulkCommand.php

```diff
@@ -0,0 +1,120 @@
+ *
+ * @author Konstantin Sirotkin <beowulfenator@gmail.com>
+ * @since 2.0.5
+ */
+class BulkCommand extends Component
+{
+    /**
+     * @var Connection
+     */
+    public $db;
+    /**
+     * @var string Default index to execute the queries on. Defaults to null meaning that index needs to be specified in every action.
+     */
+    public $index;
+    /**
+     * @var string Default type to execute the queries on. Defaults to null meaning that type needs to be specified in every action.
+     */
+    public $type;
+    /**
+     * @var array|string Actions to be executed in this bulk command, given as either an array of arrays or as one newline-delimited string.
+     * All actions except delete span two lines.
+     */
+    public $actions;
+    /**
+     * @var array Options to be appended to the query URL.
+     */
+    public $options = [];
+
+    /**
+     * Executes the bulk command.
+     * @return mixed
+     * @throws yii\base\InvalidCallException
+     */
+    public function execute()
+    {
+        //valid endpoints are /_bulk, /{index}/_bulk, and {index}/{type}/_bulk
+        if (null === $this->index && null === $this->type) {
+            $endpoint = ['_bulk'];
+        } elseif (null !== $this->index && null === $this->type) {
+            $endpoint = [$this->index, '_bulk'];
+        } elseif (null !== $this->index && null !== $this->type) {
+            $endpoint = [$this->index, $this->type, '_bulk'];
+        } else {
+            throw new InvalidCallException('Invalid endpoint: if type is defined, index must be defined too.');
+        }
+
+        if (empty($this->actions)) {
+            $body = '{}';
+        } elseif (is_array($this->actions)) {
+            $body = '';
+            foreach ($this->actions as $action) {
+                $body .= Json::encode($action)."\n";
+            }
+        } else {
+            $body = $this->actions;
+        }
+
+        return $this->db->post($endpoint, $this->options, $body);
+    }
+
+    /**
+     * Adds an action to the command. Will overwrite existing actions if they are specified as a string.
+     * @param array $action Action expressed as an array (will be encoded to JSON automatically).
+     */
+    public function addAction($line1, $line2 = null)
+    {
+        if (!is_array($this->actions)) {
+            $this->actions = [];
+        }
+
+        $this->actions[] = $line1;
+
+        if (null !== $line2) {
+            $this->actions[] = $line2;
+        }
+    }
+
+    public function addDeleteAction($index, $type, $id)
+    {
+        $actionData = [];
+        if (!empty($index)) {
+            $actionData['_index'] = $index;
+        }
+
+        if (!empty($type)) {
+            $actionData['_type'] = $type;
+        }
+
+        if (!empty($id)) {
+            $actionData['_id'] = $id;
+        }
+
+        $this->addAction(['delete' => $actionData]);
+    }
+
+    //TODO: index action
```

Comment:
No, I'm done. I think this is usable as it is, and I'm not sure we even need those commands - they are more like syntactic sugar.

Take a look at [this line](https://github.com/beowulfenator/yii2-elasticsearch/blob/30f31c3c2ceaa2971fe6ceb0272803f6b12f4b7b/ActiveRecord.php#L612) and [this line](https://github.com/beowulfenator/yii2-elasticsearch/blob/30f31c3c2ceaa2971fe6ceb0272803f6b12f4b7b/ActiveRecord.php#L663). I think this is good enough.

I would rather merge this as is.